Digipost/APIClient.swift

```diff
@@ -124,72 +125,47 @@ class APIClient : NSObject, NSURLSessionTaskDelegate, NSURLSessionDelegate, NSUR
         let rootResource = __ROOT_RESOURCE_URI__
-        let task = urlSessionJSONTask(url: rootResource, success: success) { (error) -> () in
-            if error.code == Constants.Error.Code.oAuthUnathorized {
-                self.updateRootResource(success: success, failure: failure)
-            } else {
-                failure(error: error)
-            }
+        validate(token: highestToken) {
+            let task = self.urlSessionJSONTask(url: rootResource, success: success, failure: failure)
+            task!.resume()
         }
-
-        validate(token: highestToken, thenPerformTask: task!)
     }
 
     func updateRootResource(#scope: String, success: (Dictionary<String,AnyObject>) -> Void , failure: (error: APIError) -> ()) {
         let rootResource = __ROOT_RESOURCE_URI__
         self.updateAuthorizationHeader(scope)
-        let task = urlSessionJSONTask(url: rootResource, success: success) { (error) -> () in
-            if error.code == Constants.Error.Code.oAuthUnathorized {
-                self.updateRootResource(scope: scope, success: success, failure: failure)
-            } else {
-                failure(error: error)
-            }
+        validateFullScope {
+            let task = self.urlSessionJSONTask(url: rootResource, success: success, failure: failure)
+            task?.resume()
         }
-        validateTokensThenPerformTask(task!)
     }
 
     func updateBankAccount(#uri : String, success: (Dictionary<String,AnyObject>) -> Void , failure: (error: APIError) -> ()) {
-        let task = urlSessionJSONTask(url: uri, success: success) { (error) -> () in
-            if error.code == Constants.Error.Code.oAuthUnathorized.rawValue {
-                self.updateBankAccount(uri: uri, success: success, failure: failure)
-            } else {
-                failure(error: error)
-            }
+        validateFullScope {
+            let task = self.urlSessionJSONTask(url: uri, success: success, failure: failure)
+            task?.resume()
         }
-        validateTokensThenPerformTask(task!)
     }
 
     func sendInvoideToBank(invoice: POSInvoice , success: () -> Void , failure: (error: APIError) -> ()) {
-        let task = urlSessionTask(httpMethod.post, url: invoice.sendToBankUri, success: success) { (error) -> () in
-            if error.code == Constants.Error.Code.oAuthUnathorized {
-                self.sendInvoideToBank(invoice, success: success, failure: failure)
-            } else {
-                failure(error: error)
-            }
+        validateFullScope {
+            let task = self.urlSessionTask(httpMethod.post, url: invoice.sendToBankUri, success: success, failure: failure)
+            task?.resume()
         }
-        validateTokensThenPerformTask(task!)
     }
 
     func updateReceiptsInMailboxWithDigipostAddress(digipostAddress: String, uri: String, success: (Dictionary<String,AnyObject>) -> Void , failure: (error: APIError) -> ()) {
```

Comment:
Isn't this really a generic `postJSON` function now? It does the same as `updateBankAccount`. Not sure if it matters as the code probably is more readable this way. Any reason why we accept the URL as an argument here when we don't other places..?